ext/intl/dateformat/datepatterngenerator_class.cpp

```diff
@@ -0,0 +1,141 @@
+/*
+   +----------------------------------------------------------------------+
+   | This source file is subject to version 3.01 of the PHP license,      |
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available through the world-wide-web at the following url:           |
+   | http://www.php.net/license/3_01.txt                                  |
+   | If you did not receive a copy of the PHP license and are unable to   |
+   | obtain it through the world-wide-web, please send a note to          |
+   | license@php.net so we can mail you a copy immediately.               |
+   +----------------------------------------------------------------------+
+   | Authors: Mel Dafert (mel@dafert.at)                                  |
+   +----------------------------------------------------------------------+
+*/
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "../intl_cppshims.h"
+
+#include <unicode/dtptngen.h>
+
+extern "C" {
+#define USE_DATETIMEPATTERNGENERATOR_POINTER 1
+#include "datepatterngenerator_class.h"
+#include "datepatterngenerator_arginfo.h"
+#include <zend_exceptions.h>
+#include <assert.h>
+}
+
+using icu::DateTimePatternGenerator;
+using icu::Locale;
+
+/* {{{ Global variables */
+zend_class_entry *IntlDatePatternGenerator_ce_ptr;
+zend_object_handlers IntlDatePatternGenerator_handlers;
+/* }}} */
+
+/* {{{ clone handler for IntlDatePatternGenerator */
+static zend_object *IntlDatePatternGenerator_object_clone(zend_object *object)
+{
+	IntlDatePatternGenerator_object		*dtpgo_orig,
+											*dtpgo_new;
+	zend_object								*ret_val;
+	intl_error_reset(NULL);
+
+	dtpgo_orig = php_intl_datepatterngenerator_fetch_object(object);
+	intl_error_reset(DTPATTERNGEN_ERROR_P(dtpgo_orig));
+
+	ret_val = IntlDatePatternGenerator_ce_ptr->create_object(object->ce);
+	dtpgo_new  = php_intl_datepatterngenerator_fetch_object(ret_val);
+
+	zend_objects_clone_members(&dtpgo_new->zo, &dtpgo_orig->zo);
+
+	if (dtpgo_orig->dtpg != NULL) {
+		DateTimePatternGenerator	*newDtpg;
+
+		newDtpg = dtpgo_orig->dtpg->clone();
+		if (!newDtpg) {
+			zend_string *err_msg;
+			intl_errors_set_code(DTPATTERNGEN_ERROR_P(dtpgo_orig),
+				U_MEMORY_ALLOCATION_ERROR);
+			intl_errors_set_custom_msg(DTPATTERNGEN_ERROR_P(dtpgo_orig),
+				"Could not clone IntlDatePatternGenerator", 0);
+			err_msg = intl_error_get_message(DTPATTERNGEN_ERROR_P(dtpgo_orig));
+			zend_throw_exception(NULL, ZSTR_VAL(err_msg), 0);
+			zend_string_free(err_msg);
+		} else {
+			dtpgo_new->dtpg = newDtpg;
+		}
+	} else {
+		zend_throw_exception(NULL, "Cannot clone unconstructed IntlDatePatternGenerator", 0);
+	}
+
+	return ret_val;
+}
+/* }}} */
+
+/* {{{ void IntlDatePatternGenerator_object_init(IntlDatePatternGenerator_object* to)
```

Comment:
Please remove these comments (as well as any `{{{` and `}}}`, they only exists because of historical reasons.

Reply:
Thank you - I removed them.